### ISSUE TITLE
Extract line info from all parser errors

### ISSUE THREAD

**Issue by AshesITR** [r-lib/lintr#894]
The following code extracts all errors from `gram.c`:

```r
lines <- readLines("https://raw.githubusercontent.com/wch/r-source/trunk/src/main/gram.c")
err_lines <- grep("error(_(", lines, fixed = TRUE, value = TRUE)
errors <- trimws(gsub("^.*error\\(_\\(\"(.+)\".+", "\\1", err_lines))
unique(errors)
```

Looks like the following are the unique error messages:

```
> unique(errors)                                                                                                                         
 [1] "EOF whilst reading MBCS char at line %d"                                                        
 [2] "invalid multibyte character in parser at line %d"                                               
 [3] "input buffer overflow at line %d"                                                               
 [4] "The pipe operator requires a function call as RHS"                                              
 [5] "RHS variable must be a symbol"                                                                  
 [6] "incorrect tag type at line %d"                                                                  
 [7] "allocation of source reference state failed"                                                    
 [8] "contextstack overflow"                                                                          
 [9] "repeated formal argument '%s' on line %d"                                                       
[10] "nul character not allowed (line %d)"                                                            
[11] "\\\\uxxxx sequences not supported inside backticks (line %d)"                                   
[12] "invalid \\\\u{xxxx} sequence (line %d)"                                                         
[13] "\\\\Uxxxxxxxx sequences not supported inside backticks (line %d)"                               
[14] "invalid \\\\U{xxxxxxxx} sequence (line %d)"                                                     
[15] "invalid \\\\U{xxxxxxxx} value %6x (line %d)"                                                    
[16] "invalid \\\\Uxxxxxxxx value %6x (line %d)"                                                      
[17] "mixing Unicode and octal/hex escapes in a string is not allowed"                                
[18] "string at line %d containing Unicode escapes not in this locale\\nis too long (max 10000 chars)"
[19] "malformed raw string literal at line %d"                                                        
[20] "contextstack overflow at line %d"                                                               
[21] "internal parser error at line %d"
```

_Originally posted by @renkun-ken in https://github.com/r-lib/lintr/issues/893#issuecomment-997368264_

**Comment by renkun-ken** [comment]
Thanks for the generalized solution!

I tried the following 

```
> str(lintr::lint(text = 'ok\nok\n"\\u{9999"'))                                                         
List of 1
 $ <text>:List of 8
  ..$ filename     : chr "<text>"
  ..$ line_number  : int 1
  ..$ column_number: int 1
  ..$ type         : chr "error"
  ..$ message      : chr "invalid \\u{xxxx} sequence (line 3)"
  ..$ line         : chr ""
  ..$ ranges       : NULL
  ..$ linter       : chr NA
  ..- attr(*, "class")= chr "lint"
 - attr(*, "class")= chr "lints"
```

but it seems the line number is not captured.
